src/FirmwarePlugin/APM/APMParameterMetaData.cc

```diff
@@ -78,29 +79,396 @@ QVariant APMParameterMetaData::_stringToTypedVariant(const QString& string, Fact
+                // Done loading this parameter
+                // Reset for next parameter
+                qCDebug(APMParameterMetaDataLog) << "done loading parameter";
+                rawMetaData = NULL;
+                badMetaData = false;
+                xmlState.pop();
+            } else if (elementName == "parameters") {
+                qCDebug(APMParameterMetaDataLog) << "end of parameters for category: " << currentCategory;
+                correctGroupMemberships(_vehicleTypeToParametersMap[currentCategory], groupMembers);
+                groupMembers.clear();
+                xmlState.pop();
+            } else if (elementName == "vehicles") {
+                qCDebug(APMParameterMetaDataLog) << "vehicles end here, libraries will follow";
+                xmlState.pop();
+            }
+        }
+        xml.readNext();
+    }
+
+//    foreach(const QString& vehicle, _vehicleTypeToParametersMap.keys()) {
+//        qDebug() << "dumpig metadata for behcile type " << vehicle;
+//        foreach(const QString& name, _vehicleTypeToParametersMap[vehicle].keys()) {
+//            qDebug() << "has meta data for " << name << _vehicleTypeToParametersMap[vehicle][name]->name ;
+//        }
+//    }
+}
+
+void APMParameterMetaData::correctGroupMemberships(ParameterNametoFactMetaDataMap& parameterToFactMetaDataMap,
+                                                   QMap<QString,QStringList>& groupMembers)
+{
+    foreach(const QString& groupName, groupMembers.keys()) {
+            if (groupMembers[groupName].count() == 1) {
+                foreach(const QString& parameter, groupMembers.value(groupName)) {
+                    parameterToFactMetaDataMap[parameter]->group = "others";
+                }
+            }
+        }
+}
+
+bool APMParameterMetaData::skipXMLBlock(QXmlStreamReader& xml, const QString& blockName)
+{
+    QString elementName;
+    do {
+        xml.readNext();
+        elementName = xml.name().toString();
+    } while ((elementName != blockName) && (xml.isEndElement()));
+    return !xml.isEndDocument();
+}
+
+bool APMParameterMetaData::parseParameterAttributes(QXmlStreamReader& xml, APMFactMetaDataRaw* rawMetaData)
+{
+    QString elementName = xml.name().toString();
+    QList<QPair<QString,QString> > values;
+    // as long as param doens't end
+    while (!(elementName == "param" && xml.isEndElement())) {
+        if (elementName.isEmpty()) {
+            // skip empty elements. Somehow I am getting lot of these. Dont know what to do with them.
+        } else if (elementName == "field") {
+            QString attributeName = xml.attributes().value("name").toString();
+            if ( attributeName == "Range") {
+                QString range = xml.readElementText().trimmed();
+                QStringList rangeList = range.split(' ');
+                if (rangeList.count() != 2) {
+                    qCWarning(APMParameterMetaDataLog) << "space seperator didn't work',trying 'to' separator";
```

Comment:
These should be qDebug not qWarning. This is what is currently causing the unit tests to fail in travis.